source_control/gh_keys.py

```diff
@@ -0,0 +1,216 @@
+
+class GHKeys(object):
+
+  def __init__(self, module):
+    self.module = module
+    self.action = module.params['action']
+    self.user   = module.params['user']
+    self.password = module.params['password']
+    self.title  = module.params['title']
+    self.key    = module.params['key']
+    self.key_id = module.params['key_id']
+  
+  def perform_by_action(self):
+    print self.action
+    return {
+      'list_keys'  : self.list_keys,
+      'get_key'    : self.get_key,
+      'present'    : self.add_key,
+      'absent' : self.remove_key
+    }[self.action]()
+
+  def list_keys(self):
+    if self.password is None:
+      url = GH_API_URL % "users/%s/keys" % self.user
+      response, info = fetch_url(self.module, url)
+    else:
+      url = GH_API_URL % "user/keys"
+      headers = self.get_auth_header(self.user, self.password)
+      response, info = fetch_url(self.module, url, headers=headers)
+    
+    return self.handle_response(response, info)
+ 
+  def get_key(self):
+    self.validate_fields('get_key', ['key_id', 'password'])    
+
+    url = GH_API_URL %  "user/keys/%s" % self.key_id
+    headers = self.get_auth_header(self.user, self.password)
+    response, info = fetch_url(self.module, url, headers=headers)
+
+    return self.handle_response(response, info)
+
+  def add_key(self):
+    self.validate_fields('add_key', ['title', 'key', 'password'])
+
+    f = open(self.key, 'r')
+    try:
+      content = f.readline()
+    finally:
+      f.close()
+    
+    url = GH_API_URL %  "user/keys"
+    headers = self.get_auth_header(self.user, self.password)
+    data = json.dumps({ 'title' : self.title, 'key' : content })
+    response, info = fetch_url(self.module, url, headers=headers, data=data, method='POST')
+      
+    return self.handle_response(response, info)
+
+  def remove_key(self):
+    self.validate_fields('remove_key', ['key_id', 'password'])
+
+    url = GH_API_URL %  "user/keys/%s" % self.key_id
+    headers = self.get_auth_header(self.user, self.password)
+    response, info = fetch_url(self.module, url, headers=headers, method='DELETE')
+
+    return self.handle_response(response, info)
+
+  def get_auth_header(self, user, password):
+    auth = base64.encodestring('%s:%s' % (user, password)).replace('\n', '')
+    headers = {
+      'Authorization': 'Basic %s' % auth,
+    }
+    return headers
+
+  def validate_fields(self, action, fields):
+    for field in fields:
+      if getattr(self, field) is None:
+        raise ValueError(field + " cannot be null for action [" + action + "]")
+
+  def handle_response(self, response, info):
+    if info['status'] == 200:
+      return response.read()
+    else:
+      raise RuntimeError(info['msg']) 
+
+def main():
+  module = AnsibleModule(
+    argument_spec = dict(
+      password = dict(),
+      title    = dict(),
+      key      = dict(),
+      key_id   = dict(),
+      user     = dict(required=True),
+      action   = dict(choices=['present', 'absent']),
```

Comment:
Usually ansible preferes modules handle the "state" instead of performing just "actions" or "commands". rename `action` to `state` and default it to `present`